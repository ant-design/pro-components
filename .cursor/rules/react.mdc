---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
