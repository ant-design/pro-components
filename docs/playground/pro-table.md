---
title: ProTable
atomId: ProTable
nav:
  title: Playground
  path: /playground
---

# ProTable Playground

<code src="../../demos/table/demos/dynamic-settings.tsx"  background="var(--main-bg-color)" title="属性展示"></code>
