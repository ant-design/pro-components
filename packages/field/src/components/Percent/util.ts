/** 获取展示符号 */
export function getSymbolByRealValue(realValue: number) {
  if (realValue === 0) {
    return null;
  }
  if (realValue > 0) {
    return '+';
  }
  return '-';
}

/** 获取颜色 */
export function getColorByRealValue(realValue: number /** ,color: string */) {
  if (realValue === 0) {
    return '#595959';
  }
  return realValue > 0 ? '#ff4d4f' : '#52c41a';
}

/** 获取到最后展示的数字 */
export function getRealTextWithPrecision(
  realValue: number,
  precision: number = 2,
) {
  return precision >= 0 ? realValue?.toFixed(precision) : realValue;
}

/**
 * 转化为数字
 * @copy from https://github.com/toss/es-toolkit/blob/32a183828c244d675f46810935e45dfefec81a54/src/compat/util/toNumber.ts#L19
 */
export function toNumber(value: any): number {
  if (typeof value === 'symbol' || value instanceof Symbol) {
    return NaN;
  }

  return Number(value);
}
