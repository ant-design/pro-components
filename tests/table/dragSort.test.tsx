import { MenuOutlined } from '@ant-design/icons';
import { DragSortTable } from '@ant-design/pro-table';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { act } from 'react';
import { waitForWaitTime } from '../util';

async function dragAndDrop(cell: Element) {
  fireEvent.mouseDown(cell);
  await waitForWaitTime(200);
  fireEvent.mouseMove(cell, {
    clientX: 100,
    clientY: 100,
  });
  await waitForWaitTime(200);
  fireEvent.mouseMove(cell, {
    clientX: 100,
    clientY: 10,
  });
  await waitForWaitTime(200);
  fireEvent.mouseUp(cell);
  fireEvent.mouseLeave(cell);
  await waitForWaitTime(200);
}

const height = 20;
const width = 100;
const offsetHeight = 'offsetHeight';
const offsetWidth = 'offsetWidth';
/*
  el.getBoundingClientRect mock
*/
const mockGetBoundingClientRect = (element: any, index: number) =>
  vi.spyOn(element, 'getBoundingClientRect').mockImplementation(() => ({
    bottom: 0,
    height,
    left: 0,
    right: 0,
    top: index * height,
    width,
    x: 0,
    y: index * height,
  }));

afterEach(() => {
  cleanup();
});

describe('dragSort', () => {
  const originalOffsetHeight = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    offsetHeight,
  ) as any;
  const originalOffsetWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    offsetWidth,
  ) as any;

  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, offsetHeight, {
      configurable: true,
      value: height,
    });
    Object.defineProperty(HTMLElement.prototype, offsetWidth, {
      configurable: true,
      value: width,
    });
  });

  afterAll(() => {
    Object.defineProperty(
      HTMLElement.prototype,
      offsetHeight,
      originalOffsetHeight,
    );
    Object.defineProperty(
      HTMLElement.prototype,
      offsetWidth,
      originalOffsetWidth,
    );
  });

  it('üî• [dragSort] render drag sort default handle by dragSortKey', async () => {
    const onDragSortEndFn = vi.fn();
    type DataSourceItemStruct = {
      id: string;
      name: string;
    };
    const dataSource: DataSourceItemStruct[] = [
      {
        id: '1',
        name: 'kiner',
      },
      {
        id: '2',
        name: 'WenHui Tang',
      },
    ];
    const { container } = render(
      <DragSortTable
        search={false}
        toolBarRender={false}
        size="small"
        dataSource={dataSource}
        columns={[
          {
            title: 'ÊéíÂ∫è',
            dataIndex: 'sort',
          },
          {
            title: 'Id',
            dataIndex: 'id',
          },
          {
            title: 'ÂßìÂêç',
            dataIndex: 'name',
          },
        ]}
        onDragSortEnd={(data) => {
          //@ts-ignore
          onDragSortEndFn(data[0].name);
          console.log(data);
        }}
        rowKey="id"
        dragSortKey="sort"
      />,
    );

    const draggables = container.querySelectorAll(
      '[aria-roledescription="sortable"]',
    );
    Object.setPrototypeOf(window, Window.prototype);

    draggables.forEach((draggable, index) => {
      mockGetBoundingClientRect(draggable, index);
    });

    container
      .querySelectorAll('.ant-pro-table-drag-icon')
      .forEach((dragHandle, index) => {
        mockGetBoundingClientRect(dragHandle, index);
      });

    const dragHandle = container.querySelectorAll(
      '.ant-pro-table-drag-icon',
    )[1];

    await act(() => {
      return dragAndDrop(dragHandle);
    });

    // await waitFor(() => {
    //   expect(onDragSortEndFn).toBeCalled();
    // });

    expect(container.querySelector('.dragSortDefaultHandle')).toMatchSnapshot();
  });

  it('üî• [dragSort] render drag sort custom handle by dragSortHandlerRender', async () => {
    type DataSourceItemStruct = {
      id: number;
      name: string;
    };
    const dataSource: DataSourceItemStruct[] = [
      {
        id: 1,
        name: 'kiner',
      },
    ];
    const callback = vi.fn();
    const dragHandleRender = (rowData: any, idx: any) => {
      callback(rowData.name, idx);
      return (
        <>
          <MenuOutlined
            className="dragSortCustomHandle"
            style={{ cursor: 'grab', color: 'gold' }}
          />
          {idx + 1} - {rowData.name}
        </>
      );
    };
    const { container } = render(
      <DragSortTable
        size="small"
        dataSource={dataSource}
        columns={[
          {
            title: 'Id',
            dataIndex: 'id',
          },
          {
            title: 'ÂßìÂêç',
            dataIndex: 'name',
          },
        ]}
        rowKey="id"
        dragSortKey="id"
        dragSortHandlerRender={dragHandleRender}
      />,
    );

    expect(container.querySelector('.dragSortCustomHandle')).toMatchSnapshot();
    expect(callback).toBeCalled();
    expect(callback).toHaveBeenCalledWith('kiner', 0);
  });

  it('üî• [dragSort] custom render function', async () => {
    type DataSourceItemStruct = {
      id: number;
      name: string;
    };
    const dataSource: DataSourceItemStruct[] = [
      {
        id: 1,
        name: 'kiner',
      },
      {
        id: 2,
        name: 'WenHui Tang',
      },
      {
        id: 3,
        name: 'Kiner Tang',
      },
    ];
    const dragHandleRender = (rowData: any, idx: any) => (
      <>
        <MenuOutlined
          className="dragSortCustomHandle"
          style={{ cursor: 'grab', color: 'gold' }}
        />
        {idx + 1} - {rowData.name}
      </>
    );
    const callback = vi.fn();
    const { container } = render(
      <DragSortTable
        size="small"
        dataSource={dataSource}
        columns={[
          {
            title: 'ÊéíÂ∫è',
            dataIndex: 'sort',
            render: (dom, rowData, index) => {
              callback(rowData.name, index);
              return (
                <span className="customRender">{`Ëá™ÂÆö‰πâÊéíÂ∫è[${rowData.name}-${index}]`}</span>
              );
            },
          },
          {
            title: 'Id',
            dataIndex: 'id',
          },
          {
            title: 'ÂßìÂêç',
            dataIndex: 'name',
          },
        ]}
        rowKey="id"
        dragSortKey="sort"
        dragSortHandlerRender={dragHandleRender}
      />,
    );

    expect(container.querySelector('.dragSortCustomHandle')).toMatchSnapshot();
    expect(container.querySelector('.customRender')).toMatchSnapshot();
    expect(callback).toBeCalled();
    expect(callback).toHaveBeenCalledWith('kiner', 0);
  });
});
