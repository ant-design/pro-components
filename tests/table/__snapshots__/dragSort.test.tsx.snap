// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dragSort > 🔥 [dragSort] custom render function 1`] = `
<span
  aria-label="menu"
  class="anticon anticon-menu dragSortCustomHandle"
  role="img"
  style="cursor: grab; color: gold;"
>
  <svg
    aria-hidden="true"
    data-icon="menu"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
    />
  </svg>
</span>
`;

exports[`dragSort > 🔥 [dragSort] custom render function 2`] = `
<span
  class="customRender"
>
  自定义排序[kiner-0]
</span>
`;

exports[`dragSort > 🔥 [dragSort] render drag sort custom handle by dragSortHandlerRender 1`] = `
<span
  aria-label="menu"
  class="anticon anticon-menu dragSortCustomHandle"
  role="img"
  style="cursor: grab; color: gold;"
>
  <svg
    aria-hidden="true"
    data-icon="menu"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
    />
  </svg>
</span>
`;

exports[`dragSort > 🔥 [dragSort] render drag sort default handle by dragSortKey 1`] = `null`;
